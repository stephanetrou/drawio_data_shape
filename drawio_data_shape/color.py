from enum import StrEnum


class COLOR(StrEnum):
    DEFAULT_COLOR = "#535C71"

    SOURCE = "#60A917"
    MODEL = "#1BA1E2"
    CONSUMER = "#6A00FF"

    AWS_STORAGE = "#7AA116"
    AWS_DATABASE = "#C925D1"
    AWS_ANALYTICS = "#8C4FFF"

    ORACLE = "#DA4026"
    CUBE_BLUE = "#2072B8"

    BRONZE = "#F0A30A"
    DARK_BRONZE = "#CD7F32"
    SILVER = "#C0C0C0"
    GOLD = "#FFD700"
    PLATINUM = "#E5E4E2"
    DIAMOND = "#70D1F4"
    LIGHT_DIAMOND = "#B9F2FF"
