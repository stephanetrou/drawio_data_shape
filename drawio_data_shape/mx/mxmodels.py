from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass, field, fields


@dataclass
class MxRectangle:
    x: None | float = None
    y: None | float = None
    width: None | float = None
    height: None | float = None

    def accept(self, visitor: Visitor, **kwargs):
        visitor.visitMxRectangle(self, **kwargs)


@dataclass
class MxGeometry:
    x: None | float = None
    y: None | float = None
    width: None | float = None
    height: None | float = None
    alternateBounds: None | MxRectangle = None

    def accept(self, visitor: Visitor, **kwargs):
        visitor.visitMxGeometry(self, **kwargs)


@dataclass
class UserObject:
    id: None | int = None
    cell: MxCell = None
    label: None | str = None
    tooltip: None | str = None
    link: None | str = None
    linkTarget: None | str = None

    def accept(self, visitor: Visitor, **kwargs):
        visitor.visitUserModel(self, **kwargs)


@dataclass
class MxCell:
    id: None | int = None
    value: None | str | UserObject = None
    geometry: None | MxGeometry = None
    style: None | dict[str, str | None] = None
    vertex: None | bool = False
    edge: None | bool = False
    connectable: None | bool = True
    visible: None | bool = True
    collapsed: None | bool = False
    parent: None | int = None
    source: None | int = None
    target: None | int = None
    children: None | list[int] = None
    edges: None | list[int] = None

    def accept(self, visitor: Visitor, **kwargs):
        visitor.visitMxCell(self, **kwargs)


@dataclass
class MxObject:
    id: int
    cell: None | MxCell = None
    label: None | str = None
    value: None | str = None
    tooltip: None | str = None

    def accept(self, visitor: Visitor):
        visitor.visitMxObject(self)


@dataclass
class MxGraphModel:
    root: list[MxCell | MxObject | UserObject] = field(default_factory=list)

    def append(self, item: MxCell | MxObject | UserObject):
        self.root.append(item)

    def accept(self, visitor: Visitor, **kwargs):
        visitor.visitMxGraphModel(self, **kwargs)


def has_default_value(instance, field_name):
    field_info = next(f for f in fields(instance) if f.name == field_name)
    default_value = field_info.default
    current_value = getattr(instance, field_name)
    return current_value == default_value


class Visitor(ABC):
    @abstractmethod
    def visitMxGraphModel(self, mx_graph_model: MxGraphModel, **kwargs) -> None:
        pass

    @abstractmethod
    def visitMxCell(self, mx_cell: MxCell, **kwargs) -> None:
        pass

    @abstractmethod
    def visitMxGeometry(self, mx_geometry: MxGeometry, **kwargs) -> None:
        pass

    @abstractmethod
    def visitUserModel(self, user_object: UserObject, **kwargs) -> None:
        pass

    @abstractmethod
    def visitMxRectangle(self, mx_rectangle: MxRectangle, **kwargs) -> None:
        pass

    @abstractmethod
    def visitMxObject(self, mx_object: MxObject, **kwargs) -> None:
        pass
