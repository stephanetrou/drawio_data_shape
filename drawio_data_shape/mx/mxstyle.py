from __future__ import annotations

from enum import StrEnum


class StyleName(StrEnum):
    # Alignment and Positioning
    ALIGN = "align"
    IMAGE_ALIGN = "imageAlign"
    VERTICAL_ALIGN = "verticalAlign"
    VERTICAL_LABEL_POSITION = "verticalLabelPosition"

    # Appearance
    ASPECT = "aspect"
    FILL_COLOR = "fillColor"
    FILL_OPACITY = "fillOpacity"
    IMAGE = "image"
    IMAGE_ASPECT = "imageAspect"
    LABEL_BACKGROUND_COLOR = "labelBackgroundColor"
    SKETCH = "sketch"
    STROKE_COLOR = "strokeColor"

    # Behavior
    ALLOW_ARROWS = "allowArrows"
    COLLAPSIBLE = "collapsible"
    CONNECTABLE = "connectable"
    CONTAINER = "container"
    DELETABLE = "deletable"
    EDITABLE = "editable"
    EDITABLE_CSS_RULES = "editableCssRules"
    EXPAND = "expand"
    LOCKED = "locked"
    META_EDIT = "metaEdit"
    MOVABLE = "movable"
    POINTER_EVENTS = "pointerEvents"
    RESIZABLE = "resizable"
    ROTATABLE = "rotatable"

    # Miscellaneous
    HTML = "html"
    ROUNDED = "rounded"
    SHAPE = "shape"
    WHITE_SPACE = "whiteSpace"
    TEXT = "text"
    FONT_STYLE = "fontStyle"
    FONT_FAMILY = "fontFamily"
    FONT_SIZE = "fontSize"
    FONT_COLOR = "fontColor"
    MOVABLE_LABEL = "movableLabel"
    NO_LABEL = "noLabel"
    OVERFLOW = "overflow"
    TEXT_SHADOW = "textShadow"
    SHADOW = "shadow"
    ROTATION = "rotation"
    TREE_FOLDING = "treeFolding"
    TREE_MOVING = "treeMoving"
    ENUMERATE = "enumerate"
    COMIC = "comic"
    OUTLINE_CONNECT = "outlineConnect"
    RECURSIVE_RESIZE = "recursiveResize"
    RES_ICON = "resIcon"


class StyleBuilder:
    def __init__(self):
        self.styles = {}

    def set_style(self, name: StyleName, value: None | str) -> StyleBuilder:
        self.styles[name.value] = value
        return self

    def build(self) -> dict[str, None | str]:
        return self.styles

    def align(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.ALIGN, value)

    def image_align(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.IMAGE_ALIGN, value)

    def vertical_align(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.VERTICAL_ALIGN, value)

    def vertical_label_position(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.VERTICAL_LABEL_POSITION, value)

    def aspect(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.ASPECT, value)

    def editable(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.EDITABLE, value)

    def fill_color(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.FILL_COLOR, value)

    def fill_opacity(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.FILL_OPACITY, value)

    def image(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.IMAGE, value)

    def image_aspect(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.IMAGE_ASPECT, value)

    def label_background_color(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.LABEL_BACKGROUND_COLOR, value)

    def sketch(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.SKETCH, value)

    def stroke_color(self, value: None | str) -> StyleBuilder:
        return self.set_style(StyleName.STROKE_COLOR, value)

    def allow_arrows(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.ALLOW_ARROWS, value)

    def collapsible(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.COLLAPSIBLE, value)

    def connectable(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.CONNECTABLE, value)

    def container(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.CONTAINER, value)

    def deletable(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.DELETABLE, value)

    def editable_css_rules(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.EDITABLE_CSS_RULES, value)

    def expand(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.EXPAND, value)

    def locked(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.LOCKED, value)

    def meta_edit(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.META_EDIT, value)

    def movable(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.MOVABLE, value)

    def pointer_events(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.POINTER_EVENTS, value)

    def resizable(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.RESIZABLE, value)

    def rotatable(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.ROTATABLE, value)

    def html(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.HTML, value)

    def rounded(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.ROUNDED, value)

    def shape(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.SHAPE, value)

    def white_space(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.WHITE_SPACE, value)

    def text(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.TEXT, value)

    def font_style(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.FONT_STYLE, value)

    def font_family(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.FONT_FAMILY, value)

    def font_size(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.FONT_SIZE, value)

    def font_color(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.FONT_COLOR, value)

    def movable_label(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.MOVABLE_LABEL, value)

    def no_label(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.NO_LABEL, value)

    def overflow(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.OVERFLOW, value)

    def text_shadow(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.TEXT_SHADOW, value)

    def shadow(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.SHADOW, value)

    def rotation(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.ROTATION, value)

    def tree_folding(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.TREE_FOLDING, value)

    def tree_moving(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.TREE_MOVING, value)

    def enumerate(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.ENUMERATE, value)

    def comic(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.COMIC, value)

    def outline_connect(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.OUTLINE_CONNECT, value)

    def recursive_resize(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.RECURSIVE_RESIZE, value)

    def res_icon(self, value: str) -> StyleBuilder:
        return self.set_style(StyleName.RES_ICON, value)
