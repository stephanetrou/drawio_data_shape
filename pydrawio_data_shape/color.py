from enum import StrEnum


class COLOR(StrEnum):
    SOURCE = "#60A917"
    MODEL = "#1BA1E2"
    CONSUMER = "#6A00FF"

    AWS_STORAGE = "#7AA116"
    AWS_DATABASE = "#C925D1"
    AWS_ANALYTICS = "#8C4FFF"
