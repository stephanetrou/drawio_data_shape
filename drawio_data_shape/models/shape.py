from dataclasses import dataclass

from drawio_data_shape.deflate_tools import encode
from drawio_data_shape.templating import ENVIRONMENT


@dataclass(frozen=True)
class ShapeInformation:
    name: str
    height: int
    width: int


def generate_encoded_shape(shape_information: ShapeInformation) -> str:
    template = ENVIRONMENT.get_template(f"shapes/{shape_information.name}.jinja2")
    content = template.render(shape=shape_information)

    encoded = encode(content)
    return encoded
