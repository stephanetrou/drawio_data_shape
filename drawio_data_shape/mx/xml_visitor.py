from __future__ import annotations

from dataclasses import asdict
from typing import Any
from xml.etree import ElementTree as ET

from drawio_data_shape.mx.mxmodels import (
    MxCell,
    MxGeometry,
    MxGraphModel,
    MxObject,
    MxRectangle,
    UserObject,
    Visitor,
    has_default_value,
)


class XMLVisitor(Visitor):
    _hierarchy: list[ET.Element]

    def __init__(self):
        self._hierarchy = []

    def _append_hierarchical_element(self, element: ET.Element) -> None:
        """
        Append the element to the hierarchy and set it as the current element
        """
        if self._hierarchy:
            self._hierarchy[-1].append(element)

        self._hierarchy.append(element)

    def _pop_hierarchical_element(self) -> None:
        """
        Remove the current element from the hierarchy
        """
        if len(self._hierarchy) > 1:
            self._hierarchy.pop()

    @staticmethod
    def _create_element(key: str, value=None) -> ET.Element:
        """
        Create an XML element with the given key and value
        """
        element = ET.Element(key)

        if value is not None:
            element.text = str(value)

        return element

    @staticmethod
    def _add_attributes(element: ET.Element, visited: Any) -> None:
        for key, value in asdict(visited).items():
            if has_default_value(visited, key):
                continue

            if isinstance(value, bool):
                element.attrib[key] = "1" if value else "0"
            elif isinstance(value, str | int | float):
                element.attrib[key] = str(value)

    def visitMxGraphModel(self, mx_graph_model: MxGraphModel, **kwargs) -> None:
        xml_graph_model = XMLVisitor._create_element("mxGraphModel")
        self._append_hierarchical_element(xml_graph_model)

        xml_root = XMLVisitor._create_element("root")
        self._append_hierarchical_element(xml_root)

        for item in mx_graph_model.root:
            item.accept(self)

        self._pop_hierarchical_element()
        self._pop_hierarchical_element()

    def visitMxObject(self, mx_object: MxObject, **kwargs) -> None:
        xml_object = XMLVisitor._create_element("object")
        self._append_hierarchical_element(xml_object)

        XMLVisitor._add_attributes(xml_object, mx_object)
        if mx_object.cell is not None:
            mx_object.cell.accept(self)

        self._pop_hierarchical_element()

    def visitMxCell(self, mx_cell: MxCell, **kwargs) -> None:
        xml_cell = XMLVisitor._create_element("mxCell")

        self._append_hierarchical_element(xml_cell)

        XMLVisitor._add_attributes(xml_cell, mx_cell)

        if mx_cell.style is not None:
            styles = [f"{k}={'none' if v is None else v}" for k, v in mx_cell.style.items()]
            xml_cell.attrib["style"] = ";".join(styles)

        if mx_cell.geometry is not None:
            mx_cell.geometry.accept(self, _as="geometry")

        self._pop_hierarchical_element()

    def visitMxGeometry(self, mx_geometry: MxGeometry, **kwargs) -> None:
        xml_geometry = XMLVisitor._create_element("mxGeometry")
        self._append_hierarchical_element(xml_geometry)

        XMLVisitor._add_attributes(xml_geometry, mx_geometry)

        if "_as" in kwargs:
            xml_geometry.attrib["as"] = kwargs["_as"]

        if mx_geometry.alternateBounds is not None:
            mx_geometry.alternateBounds.accept(self, _as="alternateBounds")

        self._pop_hierarchical_element()

    def visitMxRectangle(self, mx_rectangle: MxRectangle, **kwargs) -> None:
        xml_rectangle = XMLVisitor._create_element("mxRectangle")
        self._append_hierarchical_element(xml_rectangle)

        XMLVisitor._add_attributes(xml_rectangle, mx_rectangle)
        if "_as" in kwargs:
            xml_rectangle.attrib["as"] = kwargs["_as"]

        self._pop_hierarchical_element()

    def visitUserModel(self, user_object: UserObject, **kwargs) -> None:
        xml_user_object = XMLVisitor._create_element("UserObject")
        self._append_hierarchical_element(xml_user_object)

        XMLVisitor._add_attributes(xml_user_object, user_object)
        if user_object.cell is not None:
            user_object.cell.accept(self)

        self._pop_hierarchical_element()

    def xml_str(self) -> str:
        return ET.tostring(self._hierarchy[0], encoding="unicode")
