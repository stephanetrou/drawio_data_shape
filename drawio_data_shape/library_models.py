from dataclasses import dataclass

from drawio_data_shape.shape.color import COLOR


@dataclass(frozen=True)
class Datasource:
    height: int
    width: int


@dataclass(frozen=True)
class IconDatasource(Datasource):
    shape_icon: str
    shape_icon_color: COLOR
    shape_group_icon: str = "mxgraph.aws4.resourceIcon"


@dataclass(frozen=True)
class ImageDatasource(Datasource):
    image: str


class AWSThemes:
    S3 = IconDatasource(20, 20, "mxgraph.aws4.s3", COLOR.AWS_STORAGE)
    RDS = IconDatasource(20, 20, "mxgraph.aws4.rds", COLOR.AWS_DATABASE)
    REDSHIFT = IconDatasource(20, 20, "mxgraph.aws4.redshift", COLOR.AWS_DATABASE)
    DYNAMODB = IconDatasource(20, 20, "mxgraph.aws4.dynamodb", COLOR.AWS_DATABASE)
    GLUE = IconDatasource(20, 20, "mxgraph.aws4.glue", COLOR.AWS_ANALYTICS)
    ATHENA = IconDatasource(20, 20, "mxgraph.aws4.athena", COLOR.AWS_ANALYTICS)


class OnPremTheme:
    ORACLE = IconDatasource(20, 20, "mxgraph.aws4.database", COLOR.ORACLE)


class AzureThemes:
    API = ImageDatasource(12, 19, "img/lib/azure2/integration/API_Management_Services.svg")
    POWER_BI = ImageDatasource(20, 15, "img/lib/azure2/analytics/Power_BI_Embedded.svg")
    CUBE_BLUE = ImageDatasource(20, 18, "img/lib/active_directory/database_cube.svg")
    BLOB = ImageDatasource(17, 20, "img/lib/mscae/BlobBlock.svg")
    SQL_DATA_WAREHOUSE = ImageDatasource(20, 20, "img/lib/azure2/databases/SQL_Data_Warehouses.svg")
    SQL_SERVER = ImageDatasource(20, 20, "img/lib/azure2/databases/SQL_Server.svg")


@dataclass(frozen=True)
class DatasourceTemplate:
    title: str
    shape_color: COLOR
    placeholder: str
    style: Datasource
    width: int = 210
    height: int = 70
    aspect: str = "fixed"


DATASOURCES = [
    # Sources
    DatasourceTemplate("source S3", COLOR.SOURCE, "Source S3", AWSThemes.S3),
    DatasourceTemplate("source rds", COLOR.SOURCE, "Source RDS", AWSThemes.RDS),
    DatasourceTemplate("source redshift", COLOR.SOURCE, "Source Redshift", AWSThemes.REDSHIFT),
    DatasourceTemplate("source dynamodb", COLOR.SOURCE, "Source DynamoDB", AWSThemes.DYNAMODB),
    DatasourceTemplate("source glue", COLOR.SOURCE, "Source Glue", AWSThemes.GLUE),
    DatasourceTemplate("source athena", COLOR.SOURCE, "Source Athena", AWSThemes.ATHENA),
    DatasourceTemplate("Source api", COLOR.SOURCE, "Source API", AzureThemes.API),
    DatasourceTemplate("source blob storage", COLOR.SOURCE, "Source Blob Storage", AzureThemes.BLOB),
    DatasourceTemplate(
        "source sql data warehouse", COLOR.SOURCE, "Source SQL Data Warehouse", AzureThemes.SQL_DATA_WAREHOUSE
    ),
    DatasourceTemplate("source sql server", COLOR.SOURCE, "Source SQL Server", AzureThemes.SQL_SERVER),
    DatasourceTemplate("source oracle", COLOR.SOURCE, "Source Oracle", OnPremTheme.ORACLE),
    # Models
    DatasourceTemplate("model S3", COLOR.MODEL, "Model S3", AWSThemes.S3),
    DatasourceTemplate("model rds", COLOR.MODEL, "Model RDS", AWSThemes.RDS),
    DatasourceTemplate("model redshift", COLOR.MODEL, "Model Redshift", AWSThemes.REDSHIFT),
    DatasourceTemplate("model dynamodb", COLOR.MODEL, "Model DynamoDB", AWSThemes.DYNAMODB),
    # Consumers
    DatasourceTemplate("consumer powerbi", COLOR.CONSUMER, "Consumer PowerBI", AzureThemes.POWER_BI),
    DatasourceTemplate("consumer cube blue", COLOR.CONSUMER, "Cube", AzureThemes.CUBE_BLUE),
    DatasourceTemplate("consumer api", COLOR.CONSUMER, "Consumer API", AzureThemes.API),
    DatasourceTemplate("consumer sql server", COLOR.CONSUMER, "Consumer SQL Server", AzureThemes.SQL_SERVER),
]


@dataclass(frozen=True)
class MedallionTemplate:
    title: str
    shape_color: COLOR
    width: int = 16
    height: int = 16
    aspect: str = "fixed"


MEDALLIONS = [
    MedallionTemplate("Bronze", COLOR.BRONZE),
    # MedallionTemplate("Dark Bronze", COLOR.DARK_BRONZE),
    MedallionTemplate("Silver", COLOR.SILVER),
    MedallionTemplate("Gold", COLOR.GOLD),
    # MedallionTemplate("Platinum", COLOR.PLATINUM),
    MedallionTemplate("Diamond", COLOR.DIAMOND),
    # MedallionTemplate("Light Diamond", COLOR.LIGHT_DIAMOND),
]
