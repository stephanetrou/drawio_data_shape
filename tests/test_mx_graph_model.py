from approvaltests import verify_xml

from drawio_data_shape.library_models import AWSThemes, DatasourceTemplate
from drawio_data_shape.mx.mxmodels import MxCell, MxGeometry, MxGraphModel, MxRectangle, UserObject
from drawio_data_shape.mx.mxstyle import StyleBuilder
from drawio_data_shape.mx.xml_visitor import XMLVisitor
from drawio_data_shape.shape.color import COLOR
from drawio_data_shape.shape.datasource_container import DatasourceContainer


def test_mx_cell_empty():
    visitor = XMLVisitor()
    visitor.visitMxCell(MxCell(0))

    pretty_xml = visitor.xml_str()
    verify_xml(pretty_xml)


def test_mx_cell_parent():
    visitor = XMLVisitor()
    visitor.visitMxCell(MxCell(1, parent=0))

    pretty_xml = visitor.xml_str()
    verify_xml(pretty_xml)


def test_mx_geometry():
    visitor = XMLVisitor()
    visitor.visitMxGeometry(MxGeometry(height=12, width=15))

    pretty_xml = visitor.xml_str()
    verify_xml(pretty_xml)


def test_mx_rectangle():
    visitor = XMLVisitor()
    visitor.visitMxRectangle(MxRectangle(x=10, y=11, width=12, height=13))

    pretty_xml = visitor.xml_str()
    verify_xml(pretty_xml)


def test_mx_cell_with_geometry():
    visitor = XMLVisitor()
    visitor.visitMxCell(MxCell(2, geometry=MxGeometry(height=12, width=15)))

    pretty_xml = visitor.xml_str()
    verify_xml(pretty_xml)


def test_mx_cell_with_geometry_and_alternate_bounds():
    visitor = XMLVisitor()
    visitor.visitMxCell(MxCell(2, geometry=MxGeometry(height=12, width=15, alternateBounds=MxRectangle(x=10, y=11))))

    pretty_xml = visitor.xml_str()
    verify_xml(pretty_xml)


def test_user_object():
    visitor = XMLVisitor()
    visitor.visitUserModel(UserObject(id=1, label="Hello"))

    pretty_xml = visitor.xml_str()
    verify_xml(pretty_xml)


def test_user_object_with_mx_cell():
    visitor = XMLVisitor()
    visitor.visitUserModel(UserObject(id=1, label="Hello", cell=MxCell(2)))

    pretty_xml = visitor.xml_str()
    verify_xml(pretty_xml)


def test_simple_shape():
    graph_model = MxGraphModel()

    graph_model.append(MxCell(0))

    graph_model.append(MxCell(1, parent=0))

    style = (
        StyleBuilder()
        .rounded("1")
        .white_space("wrap")
        .html("1")
        .font_family("Verdana")
        .font_size("9")
        .stroke_color(None)
        .fill_color("#535C71")
        .font_color("#FFFFFF")
        .build()
    )

    geometry = MxGeometry(height=12, width=26)

    graph_model.append(MxCell(2, value="Hi !!", geometry=geometry, style=style, vertex=True, parent=1))

    xml_visitor = XMLVisitor()
    xml_visitor.visitMxGraphModel(graph_model)
    pretty_xml = xml_visitor.xml_str()

    verify_xml(pretty_xml)


def test_datasource_shape():
    template_detail = DatasourceTemplate("source S3", COLOR.SOURCE, "Source S3", AWSThemes.S3, "S3")

    container = DatasourceContainer()
    graph_model = container.build(template_detail)

    xml_visitor = XMLVisitor()
    xml_visitor.visitMxGraphModel(graph_model)
    pretty_xml = xml_visitor.xml_str()

    verify_xml(pretty_xml)
