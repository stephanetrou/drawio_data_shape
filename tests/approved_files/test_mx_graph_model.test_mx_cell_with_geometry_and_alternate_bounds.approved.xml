<?xml version="1.0" ?>
<mxCell id="2">
	<mxGeometry width="15" height="12" as="geometry">
		<mxRectangle x="10" y="11" as="alternateBounds"/>
	</mxGeometry>
</mxCell>
